### PR TITLE
feat: add gojq integration for custom field selection (#37)

### LINKED ISSUE #37: feat: add custom field selection with gojq integration
## Summary
Integrate gojq into gh-helper to enable custom field selection and flexible data processing for all commands.

### 1. Field Selection with --fields
Add a `--fields` flag to select specific fields from the output:

```bash
# Select specific fields
gh-helper issues show 248 --fields "title,body,createdAt"
gh-helper reviews fetch 300 --fields "reviews.author,reviews.state,reviews.submittedAt"
```

### 2. Custom Queries with --query
Add a `--query` flag to run custom jq queries on the output:

```bash
# Custom formatting
gh-helper issues show 248 --query '.title + " - " + .state'

# Complex filtering
gh-helper reviews fetch 300 --query '.reviews[] | select(.state == "APPROVED")'
```

### Implementation Approach:
1. Add gojq as a dependency (already used by github-schema-go)
2. Process the command output through gojq before final serialization
3. Apply queries after data is fetched but before YAML/JSON output

### Benefits:
- Eliminates need for external `gojq --yaml-input` piping
- Provides consistent interface across all commands
- Enables complex data transformations without additional tools
- Reduces output size by selecting only needed fields

### Considerations:
- Maintain backward compatibility (no query = full output)
- Provide helpful error messages for invalid jq expressions
- Document common query patterns in README

## Examples of Use Cases:
- Extract only unresolved review threads
- Format output for specific reporting needs
- Filter results based on complex conditions
- Transform data structure for integration with other tools

### PR DESCRIPTION
**Pull request by apstndb**
## Summary

This PR implements gojq integration for gh-helper, enabling custom field selection and flexible data processing through jq queries on command outputs.

Closes #37

### 1. Core Integration
- Added global `--jq` flag to all commands for applying jq queries to output
- Integrated `github.com/apstndb/go-jq-yamlformat` for efficient jq processing
- Supports both YAML and JSON output formats with jq filtering

### 2. Refactoring
- Refactored `output_format.go` to use `go-yamlformat` directly
- Created `EncodeOutputWithCmd` function for consistent output handling across all commands
- Updated all commands to use the new encoding pattern

### 3. Dependencies Added
- `github.com/apstndb/go-yamlformat` - Consistent YAML/JSON formatting
- `github.com/itchyny/gojq` - Pure Go implementation of jq
- `github.com/apstndb/go-jq-yamlformat` - Integration layer between gojq and yamlformat

## Usage Examples

```bash
# Extract specific fields
gh-helper issues show 37 --jq '.issueShow.issue.title'
# Output: "feat: add custom field selection with gojq integration"

# Build custom objects  
gh-helper issues show 37 --jq '.issueShow.issue | {title, state, labels}'
# Output: 
# labels:
# - enhancement
# - priority/medium
# state: OPEN
# title: "feat: add custom field selection with gojq integration"

# Filter arrays
gh-helper reviews fetch 306 --jq '.reviews[] | select(.state == "APPROVED")'

# Extract node IDs
gh-helper node-id issue 37 --jq '.nodeId.id'
# Output: I_kwDOO-Obpc68vcdL

# Complex queries with sub-issues
gh-helper issues show 248 --include-sub --jq '.issueShow.issue.subIssues.nodes[] | select(.state == "OPEN") | .title'
```

## Key Benefits

1. **No External Dependencies**: Users no longer need to pipe output to external jq
2. **Consistent Output**: jq queries work identically with both YAML and JSON formats
3. **Performance**: Direct integration is more efficient than piping
4. **Error Handling**: Better error messages for invalid jq expressions
5. **Timeout Protection**: 30-second timeout prevents hanging on complex queries

## Testing

- ✅ All existing tests pass
- ✅ Manual testing confirms jq queries work correctly across all commands
- ✅ Tested with complex nested queries and filters

## Documentation

Updated README.md with comprehensive examples and usage instructions for the new jq functionality.

🤖 Generated with [Claude Code](https://claude.ai/code)

Co-Authored-By: Claude <noreply@anthropic.com>

### CODE REVIEW

File: gh-helper/output_format.go [hunk 1]
```diff
@@ -57,30 +35,93 @@ func ResolveFormat(cmd *cobra.Command) OutputFormat {
...
+
+	switch format {
+	case FormatJSON:
+		encoder := yamlformat.NewJSONEncoder(w)
+		return encoder.Encode(data)
+	default: // YAML and others
+		encoder := yamlformat.NewEncoder(w)
+		return encoder.Encode(data)
+	}
 }
```
[the remark this answers is not in the document]
Comment (apstndb) [reply gh-helper/output_format.go:66]: Added comprehensive unit tests for the jq functionality in output_format_test.go. The tests cover various scenarios including different output formats, valid/invalid queries, and edge cases.

Fixed in commit 2d0d714.

File: README.md [hunk 2]
```diff
@@ -48,8 +49,42 @@ gh-helper issues create --title "Subtask" --body "Details" --parent 123
...
+gh-helper issues show 37 --jq '.issueShow.issue.title'
+
+# Build custom objects
+gh-helper issues show 37 --jq '.issueShow.issue | {title, state, labels}'
+
+# Filter arrays
+gh-helper reviews fetch 306 --jq '.reviews[] | select(.state == "APPROVED")'
+
+# Transform output
+gh-helper issues list --jq '.issues[].title' | sort
```
[the remark this answers is not in the document]
Comment (apstndb) [reply README.md]: Fixed the example to use an existing command. Changed from non-existent 'issues list' to 'reviews fetch' with author extraction.

Fixed in commit 2d0d714.

File: README.md [hunk 3]
```diff
@@ -48,8 +49,42 @@ gh-helper issues create --title "Subtask" --body "Details" --parent 123
...
+gh-helper issues show 37 --jq '.issueShow.issue.title'
+
+# Build custom objects
+gh-helper issues show 37 --jq '.issueShow.issue | {title, state, labels}'
+
+# Filter arrays
+gh-helper reviews fetch 306 --jq '.reviews[] | select(.state == "APPROVED")'
+
+# Transform output
+gh-helper reviews fetch 306 --jq '.reviews[].author.login' | sort | uniq
```
[the remark this answers is not in the document]
Comment (apstndb) [reply README.md]: Fixed by removing the shell pipe example to avoid confusion

Fixed in commit eec0b86.

File: gh-helper/output_format.go [hunk 4]
```diff
@@ -57,30 +35,93 @@ func ResolveFormat(cmd *cobra.Command) OutputFormat {
...
+	oldCmd := currentCommand
+	currentCommand = cmd
+	defer func() { currentCommand = oldCmd }()
+
+	format := ResolveFormat(cmd)
+	return EncodeOutput(os.Stdout, format, data)
+}
+
+// currentCommand is used to pass command context to EncodeOutput
+var currentCommand *cobra.Command
```
[the remark this answers is not in the document]
Comment (apstndb) [reply gh-helper/output_format.go:83]: Fixed by defining jqQueryTimeout as a named constant

Fixed in commit eec0b86.

File: gh-helper/output_format.go [hunk 5]
```diff
@@ -57,30 +35,93 @@ func ResolveFormat(cmd *cobra.Command) OutputFormat {
...
+	var yf yamlformat.Format
+	switch format {
+	case FormatJSON:
+		yf = yamlformat.FormatJSON
+	default:
+		yf = yamlformat.FormatYAML
+	}
+
+	// Create context with timeout
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
```
[the remark this answers is not in the document]
Comment (apstndb) [reply gh-helper/output_format.go]: The global variable is used to pass command context for the jq query flag. While not ideal, refactoring this would be a larger architectural change beyond the scope of this PR. I've noted this for future improvement.

File: gh-helper/reviews_unified.go [hunk 6]
```diff
@@ -150,20 +150,21 @@ func fetchReviews(cmd *cobra.Command, args []string) error {
...
 	// Full data output
+	format := ResolveFormat(cmd)
 	if format == FormatJSON {
-		return EncodeOutput(os.Stdout, format, data)
+		return EncodeOutputWithCmd(cmd, data)
 	}
 	
 	// Use specialized output function for better structure  
-	return outputFetch(data, includeReviewBodies, includeThreads, format)
+	return outputFetch(cmd, data, includeReviewBodies, includeThreads)
```
[the remark this answers is not in the document]
Comment (apstndb) [reply gh-helper/reviews_unified.go]: Fixed redundant ResolveFormat calls

Fixed in commit 4686d3b.

File: gh-helper/output_format.go [hunk 7]
```diff
@@ -57,30 +38,79 @@ func ResolveFormat(cmd *cobra.Command) OutputFormat {
...
 
-// Marshal marshals data using the specified format
-func (f OutputFormat) Marshal(data interface{}) ([]byte, error) {
-	options := []yaml.EncodeOption{yaml.UseJSONMarshaler()}
-	if f == FormatJSON {
-		options = append(options, yaml.JSON())
+// EncodeOutputWithCmd encodes data with optional jq query from command
+func EncodeOutputWithCmd(cmd *cobra.Command, data interface{}) error {
+	format := ResolveFormat(cmd)
+	jqQuery, _ := cmd.Root().Flags().GetString("jq")
```
[the remark this answers is not in the document]
Comment (apstndb) [reply gh-helper/output_format.go:67]: Added a comment explaining why the error is intentionally ignored. The jq flag is guaranteed to exist since it's registered in rootCmd.

Fixed in commit 1d5342e.

File: gh-helper/reviews_unified.go [hunk 8]
```diff
@@ -98,7 +95,10 @@ func fetchReviews(cmd *cobra.Command, args []string) error {
 	
 	// Adjust flags for thread-focused modes
 	if listThreads || threadsOnly {
-		format = FormatJSON // Force JSON for thread-focused modes
+		// Force JSON for thread-focused modes by setting the flag
+		if err := cmd.Flags().Set("format", "json"); err != nil {
+			return fmt.Errorf("failed to set format flag: %w", err)
+		}
```
[the remark this answers is not in the document]
Comment (apstndb) [reply gh-helper/reviews_unified.go]: Added a detailed comment explaining why the flag is programmatically set and how it interacts with the centralized output handling.

Fixed in commit 1d5342e.

File: gh-helper/reviews_unified.go [hunk 9]
```diff
@@ -150,20 +153,21 @@ func fetchReviews(cmd *cobra.Command, args []string) error {
...
+	format := ResolveFormat(cmd)
 	if format == FormatJSON {
-		return EncodeOutput(os.Stdout, format, data)
+		return EncodeOutputWithCmd(cmd, data)
 	}
 	
-	// Use specialized output function for better structure  
-	return outputFetch(data, includeReviewBodies, includeThreads, format)
+	// Use specialized output function for better YAML structure  
+	return outputFetch(cmd, data, includeReviewBodies, includeThreads)
```
[the remark this answers is not in the document]
Comment (apstndb) [reply gh-helper/reviews_unified.go]: Fixed the data structure inconsistency. Now both JSON and YAML outputs use the same structured format from outputFetch, ensuring consistent jq queries across output formats.

Fixed in commit 5da22ee.

File: gh-helper/output_format.go [hunk 10]
```diff
@@ -57,30 +38,81 @@ func ResolveFormat(cmd *cobra.Command) OutputFormat {
...
+	var yf yamlformat.Format
+	switch format {
+	case FormatJSON:
+		yf = yamlformat.FormatJSON
+	default:
+		yf = yamlformat.FormatYAML
+	}
+
+	// Create context with timeout
+	ctx, cancel := context.WithTimeout(context.Background(), jqQueryTimeout)
```
[the remark this answers is not in the document]
Comment (apstndb) [reply gh-helper/output_format.go]: Implemented context propagation as suggested. EncodeOutputWithJQ now receives the command context, enabling proper cancellation handling when the command is interrupted.

Fixed in commit 5da22ee.

File: gh-helper/output_format.go [hunk 11]
```diff
@@ -57,30 +38,81 @@ func ResolveFormat(cmd *cobra.Command) OutputFormat {
...
+	return EncodeOutput(os.Stdout, format, data)
 }
 
-// Unmarshal unmarshals data using consistent options (format-agnostic)
+
+// EncodeOutputWithJQ encodes data with optional jq query filtering
+func EncodeOutputWithJQ(ctx context.Context, w io.Writer, format OutputFormat, data interface{}, jqQuery string) error {
+	// If no jq query provided, encode normally
+	if jqQuery == "" {
+		return EncodeOutput(w, format, data)
```
[the remark this answers is not in the document]
Comment (apstndb) [reply gh-helper/output_format.go]: Removed the redundant empty jqQuery check as suggested. This improves separation of concerns and makes the function's contract clearer.

Fixed in commit 8aa3d7a.

File: gh-helper/output_format_test.go [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+	// Check for context deadline exceeded error
+	// Note: Ideally we would only check errors.Is(err, context.DeadlineExceeded),
+	// but go-jq-yamlformat currently returns "execution timeout after Xs" without
+	// properly wrapping the context.DeadlineExceeded error. This is a known issue
+	// that should be fixed in the library.
+	if !errors.Is(err, context.DeadlineExceeded) {
+		// Fallback check for the specific error message from go-jq-yamlformat
+		if !strings.Contains(err.Error(), "execution timeout after") {
+			t.Errorf("Expected context.DeadlineExceeded or 'execution timeout after', got: %v", err)
+		}
```
[the remark this answers is not in the document]
Comment (apstndb) [reply gh-helper/output_format_test.go:203]: The test already follows the suggested pattern - it first checks with errors.Is(err, context.DeadlineExceeded) and only falls back to string checking if that fails. This is necessary because go-jq-yamlformat currently doesn't properly wrap the error. I've filed an issue with the library: https://github.com/apstndb/go-jq-yamlformat/issues/3

File: gh-helper/output_format.go [hunk 13]
```diff
@@ -57,30 +38,76 @@ func ResolveFormat(cmd *cobra.Command) OutputFormat {
...
+	format := ResolveFormat(cmd)
+	// GetString error is intentionally ignored as the jq flag is guaranteed to exist
+	// (registered in rootCmd) and will return empty string if not set
+	jqQuery, _ := cmd.Root().Flags().GetString("jq")
+	
+	if jqQuery != "" {
+		return EncodeOutputWithJQ(cmd.Context(), os.Stdout, format, data, jqQuery)
+	}
+	
+	return EncodeOutput(os.Stdout, format, data)
```
[the remark this answers is not in the document]
Comment (apstndb) [reply gh-helper/output_format.go]: Fixed as suggested. Using cmd.OutOrStdout() now for better testability.

Fixed in commit 7095760e9e9d085c21866ef41fdd2470c7b7e9b2.
